_source/_posts/2022-04-05-state-of-ffi-java.md

```diff
@@ -0,0 +1,280 @@
+
+### Benchmark
+
+Let's run some [Java Microbenchmark Harness (JMH)](https://github.com/openjdk/jmh) benchmarks to compare the performance of JNI and Panama API. We will use the `getpid` function from the standard C `unistd` header for the comparison. We will call the API using JNI and Panama APIs and compare the performance. I'm running the benchmark on Linux with OpenJDK 19 early access build for Panama (`openjdk 19-panama 2022-09-20`).
+
+Here is the code for JNI, which uses the prebuilt [JavaCPP](https://github.com/bytedeco/javacpp) library to call the `getpid` function. We don't have to write all the manual C binding code and rituals as the JavaCPP library already does it.
+
+```java
+@Benchmark
+public int JNILinux() {
+    return org.bytedeco.javacpp.linux.getpid();
+}
+
+
+@Benchmark
+public int JNIMac() {
+    return org.bytedeco.javacpp.macosx.getpid();
+}
+```
+
+Here is the code for Panama, which uses the Foreign Linker API to call the `getpid` function manually.
+
+```java
+// get System linker
+private static final CLinker linker = CLinker.systemCLinker();
+// predefine symbols and method handle info
+private static final NativeSymbol nativeSymbol = linker.lookup("getpid").get();
+private static final MethodHandle getPidMH = linker.downcallHandle(
+        nativeSymbol,
+        FunctionDescriptor.of(ValueLayout.OfInt.JAVA_INT));
+
+@Benchmark
+public int panamaDowncall() throws Throwable {
+    return (int) getPidMH.invokeExact();
+}
+```
+
+With jextract, you can further simplify it by generating a Java API for the header file with the below command:
+
+```bash
+# Linux
+export C_INCLUDE=/usr/include
+# macOS
+export C_INCLUDE=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
+
+jextract --source -d generated/src/main/java -t org.unix -I $C_INCLUDE $C_INCLUDE/unistd.h
+```
+
+And here is the code for calling the API generated by jextract
+
+```java
+@Benchmark
+public int panamaJExtract() {
+    return org.unix.unistd_h.getpid();
+}
+```
+
+This is a sample result.
+
+```text
+Benchmark                    Mode  Cnt   Score   Error  Units
+FFIBenchmark.JNI             avgt   40  50.221 ± 0.512  ns/op
+FFIBenchmark.panamaDowncall  avgt   40  49.382 ± 0.701  ns/op
+FFIBenchmark.panamaJExtract  avgt   40  49.946 ± 0.721  ns/op
+```
+
+It seems like using Panama API is slightly faster than JNI, which is at the incubator stage, so I'm expecting this to become better when stable.
+
+If you would like to run the benchmarks yourself, follow the instructions in the readme file on [the source repository](https://github.com/deepu105/Java-FFI-benchmarks)
+
+## So, are we there yet?
+
+The current state of project Panama, as of JDK 18, is as follows.
+
+- Still incubating
+- Can already work with languages that have C interop like C/C++, Fortran, Rust, etc
+- Performance on par or better than JNI. Hopefully, this will be improved further
```

Comment:
```suggestion
- Performance on par or better than JNI. Hopefully, this will be improved further.
```
I'm adding periods here, where the bullets are sentences. Bending the rules a little for a more consistent appearance.